### ISSUE TITLE
Add SerialSubscription

### ISSUE THREAD

**Issue by sir-boformer** [ReactiveX/rxdart#183]
`CompositeSubscription`
------

Basically a bag for multiple subscription (with `add` and `remove` methods) that can be unsubscribed at once: http://reactivex.io/RxJava/javadoc/io/reactivex/disposables/CompositeDisposable.html

```dart
final s = new CompositeSubscription();
s.add(firstStream.listen(...));
s.add(secondStream.listen(...));
//...
s.cancel(); // cancels all
```
It's very useful when you listen to multiple streams and want to dispose them all at once (e.g. when a widget in Flutter is disposed).

Danger of this one is that developers add thousands of subscriptions (e.g. for a repeatable task). In RxJava at least the "bag" is never emptied, even if the streams already emptied. 

`SerialSubscription`
---------

 A container that keeps track of a single subscription, and you can replace the subscription with another one (`set` method), which cancels the previous subscription: http://reactivex.io/RxJava/javadoc/io/reactivex/disposables/SerialDisposable.html

```dart
final s = new SerialSubscription();
s.set(firstStream.listen(...));
// ...
s.set(secondStream.listen(...)); // cancel firstStream subscription
// ...
s.cancel(); // cancel tracked subscription (if any)
```

Basically a shorthand for:

```dart
Subscription s;

s = firstStream.listen(...);
// ...
s?.cancel();
s = secondStream.listen(...);
// ...
s?.cancel();
```

I love this one. It ensures that subscriptions never get lost when you replace them. And you can also add them to `SerialSubscription`s.

Add `subscription` argument to `doOnListen` operator
----------

Use case: Easily add/set a subscription without breaking the flow:

```dart
final s = new CompositeSubscription();

observable
  .doOnListen(s.add)
  .listen(...);
```

**Comment by sir-boformer** [comment]
What do you think? If it makes sense, I may take up the task.

**Comment by brianegan** [comment]
Hey there :) Yah, we've had this request a couple of times. I think we've always just said "You can do it manually." E.g. CompositeSubscription could be done like so `List<StreamSubscription> subs = []; subs.forEach((sub) => sub.cancel());`). But maybe it's not the best answer since this question has been asked a few times now.

I'd be open to this contribution. If you take it up, could you please ensure you write tests for each? Thanks!

**Comment by brianegan** [comment]
CompositeSubscription now added. Changing the title.

**Comment by shtse8** [comment]
how about `SerialSubscription`?
